src/Core/KeyValueStorage/Type/DBKeyValueStorage.php

```diff
@@ -0,0 +1,117 @@
+<?php
+/**
+ * @copyright Copyright (C) 2010-2022, the Friendica project
+ *
+ * @license GNU AGPL version 3 or any later version
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ */
+
+namespace Friendica\Core\KeyValueStorage\Type;
+
+use Friendica\Core\Config\Util\ValueConversion;
+use Friendica\Core\KeyValueStorage\Exceptions\KeyValueStoragePersistenceException;
+use Friendica\Database\Database;
+
+/**
+ * A Key-Value storage provider with DB as persistence layer
+ */
+class DBKeyValueStorage extends AbstractKeyValueStorage
+{
+	const DB_KEY_VALUE_TABLE = 'key-value';
+
+	/** @var Database */
+	protected $database;
+
+	public function __construct(Database $database)
+	{
+		$this->database = $database;
+	}
+
+	/** {@inheritDoc} */
+	public function offsetExists($offset): bool
+	{
+		try {
+			return $this->database->exists(self::DB_KEY_VALUE_TABLE, ['k' => $offset]);
+		} catch (\Exception $exception) {
+			throw new KeyValueStoragePersistenceException(sprintf('Cannot check storage with key %s', $offset), $exception);
+		}
+	}
+
+	/** {@inheritDoc} */
+	#[\ReturnTypeWillChange]
+	public function offsetGet($offset)
+	{
+		try {
+			$result = $this->database->selectFirst(self::DB_KEY_VALUE_TABLE, ['v'], ['k' => $offset]);
+
+			if ($this->database->isResult($result)) {
+				$value = ValueConversion::toConfigValue($result['v']);
+
+				// just return it in case it is set
+				if (isset($value)) {
+					return $value;
+				}
+			}
+		} catch (\Exception $exception) {
+			throw new KeyValueStoragePersistenceException(sprintf('Cannot get value for key %s', $offset), $exception);
+		}
+
+		return null;
+	}
+
+	/** {@inheritDoc} */
+	#[\ReturnTypeWillChange]
+	public function offsetSet($offset, $value)
+	{
+		try {
+			// We store our setting values in a string variable.
+			// So we have to do the conversion here so that the compare below works.
+			// The exception are array values.
+			$compare_value = (!is_array($value) ? (string)$value : $value);
+			$stored_value  = $this->get($offset);
+
+			if (isset($stored_value) && ($stored_value === $compare_value)) {
+				return;
+			}
+
+			$dbValue = ValueConversion::toDbValue($value);
+
+			$return = $this->database->update(self::DB_KEY_VALUE_TABLE, [
```

Comment:
Shouldn’t this be an `insert(…, INSERT_UPDATE)` call?

Reply:
It's copied from the `Config` Repository :) I'm open to this change. What's @annando opinion?

Reply:
I’m just worried this wouldn’t save new keys.

Reply:
I can say that it works very well :)
- Unit tests worked
- Local node worked
- Friendica.philipp.info worked

Reply:
Yes but you’re riding on existing keys, aren’t you?

Reply:
nope, at least the unit tests are setting new values, so does my local node after first time saving at `admin\site`

Reply:
Sounds like `update` is performing some kind of magic then.

Reply:
Yeah, see the last parameter. This is is some "insert if not existing" functionality.

Reply:
So `update()` is performing INSERT and `insert` is performing UPDATE, I feel like we could combine both. 🤔

Reply:
No, there are fine differences in it. For example the functionality in `update' can also be called with a second value array that contains the previous data. Then only the changed values will be updated and no update is executed when nothing changed.

I had thought about doing it the one or the other way, but haven't come to a final decision here.

Reply:
My remark was mostly done in jest, but what you said indicates we’re already supporting dirty/clean updates.